### DISCUSSION TITLE
op-stack support state overrides

### DISCUSSION THREAD

**Discussion by leovigna** [wevm/viem#2689] [Idea / Feature Request]
## Problem
The following op-stack actions do **not** support state overrides:
* `depositTransaction`
* `estimateDepositTransactionGas` (called by the above for gas estimation)

This hinders use cases with smart wallets where the gas cost of the transaction might be ultimately be paid by a bundler. With state overrides, we could set the balance of the smart wallet to say 1 ETH and then estimate gas properly.

## Use Case
Using op-stack extension to bridge eth using smart wallets & paymasters:
* Send 0.001 ETH to smart wallet on L1
* Bridge 0.001 ETH to L2 using viem account abstraction / permissionless to send a UserOp

This **fails** at the gas estimation stage of the UserOp when bridging the full balance of the smart wallet despite it being technically possible. This is because the gas estimation just assumes the call is being paid for by the smart wallet address.

## Solution
It would be nice for the above actions to align with the `estimateContractGas` & `estimateGas` interfaces to support state overrides.

**Comment by leovigna** [comment]
Also, if anyone is having the same issue as me for smart wallets specifically, setting the `gas` parameter in `depositTransaction` to `null` is sufficient as the gas estimation is can just be skipped for the UserOp (callGasLimit can just be set to 0)

**Comment by songkeys** [comment]
I have the same issue, but my use case is that I want to estimate the fee on the frontend to guide users to deduct the gas fee from the value instead of paying their maximum balance. The failed contract call to `eth_estimateGas` if balance is not enough seems like a paradox to this case.

@jxom, I can work on this PR. For instance, we could add a `stateOverrides` parameter to these functions to pass it to the underlying `estimateGas` function. What do you think about this idea?

**Comment by danielsimao** [comment]
Unrelated question (I raised this several times in the past)
Does viem/op support ERC20? The old document [(link)](https://www.opviem.sh/) has several references to handling ERC20's, but once it was migrated to viem, seems to have been left out.

Could you enlighten me here? Sorry for the off-topic question.

**Reply by songkeys** [reply]
I believe you are talking about this issue: https://github.com/wevm/viem/discussions/1646

Since viem's chain definitions currently only include the `portal` contract, and the `depositERC20To` method requires the `l1StandardBridge` contract, you can't fully migrate to viem just yet. I'm experiencing the same situation and am using `op-viem` along with `viem/opstack` in a mixed manner.

Also ref: https://github.com/wevm/viem/pull/1479#discussion_r1449669894
